### ISSUE TITLE
Are you actually using the learned intrinsic reward for the agent?

### ISSUE THREAD

**Issue by ferreirafabio** [jcwleo/curiosity-driven-exploration-pytorch#9]
Hi,

I can only see that you optimize the intrinsic loss in your code. Can you point me to the line where you add the intrinsic rewards to the actual environment/extrinsic rewards?

In some areas of your code I can see comments like 
`# total reward = int reward`
which would, according to the original paper, be wrong, no?  

Thank you.

**Comment by ruoshiliu** [comment]
Also new to the repo, but here the loss is composed of both intrinsic and extrinsic reward:
https://github.com/jcwleo/curiosity-driven-exploration-pytorch/blob/bacbefdfbdbc4c4382ab67147c9c8410305a4978/agents.py#L144

**Comment by ferreirafabio** [comment]
Thanks @ruoshiliu. Yes, I saw the loss. But in addition to optimizing the loss you also need to use the intrinsic rewards (which is the result from optimizing its loss) for the agent as stated in the paper. Only optimizing the loss is not equivalent to using the intrinsic reward as an outcome of optimizing its loss.

**Comment by ruoshiliu** [comment]
@ferreirafabio What do you mean by "use the intrinsic rewards"? Can you point out which section in the paper stated that?

**Comment by ferreirafabio** [comment]
By that I mean reward = extrinsic reward + intrinsic reward. From the paper: 

![31B16992-8338-4C37-A1AE-6983E1EB9AF1](https://user-images.githubusercontent.com/4440904/109923743-852f4f00-7cbf-11eb-80fd-26fad356d414.jpeg)

I now realize that the paper says the extrinsic reward can be optional. Wondering what is „usually“ used (with or without extrinsic reward) when peers use ICM as a baseline.

**Comment by ruoshiliu** [comment]
Thank you for the clarification. Let me make sure I understand your question. What you are saying is the code (referenced above) tries to minimize the loss function by **maximizing the extrinsic reward and minimizing the intrinsic reward**. The correct implementation should reflect the equation (7) below in which 
<img src="https://latex.codecogs.com/svg.latex?\Large&space;r_t=r^i_t+r^e_t" title="equation" />
In other words, the correct implementation should find the policy p that maximizes both intrinsic and extrinsic reward and parameters for inverse model and forward model that minimizes L_I and L_F.

Did I interpret your question correctly?

<img width="420" alt="Screen Shot 2021-03-04 at 16 40 59" src="https://user-images.githubusercontent.com/25470644/110034407-673a0c80-7d08-11eb-805c-4d9ed8239a45.png">

**Comment by ferreirafabio** [comment]
Yes

**Comment by ruoshiliu** [comment]
I think perhaps the best way to figure out what exactly is the right thing to do is to go back to the original tensorflow repo. https://github.com/pathak22/noreward-rl
